passes/hierarchy/hierarchy.cc

```diff
@@ -710,6 +711,422 @@ RTLIL::Wire *find_implicit_port_wire(Module *module, Cell *cell, const std::stri
+			return bind_path(design, counts, path, binding, top_mod);
+		}
+	}
+
+	// If no luck, try to interpret the name as a hierarchical reference
+	// starting at the top level.
+	std::vector<RTLIL::Binding::Path> paths = binding.find_tl_cells(design);
+	if (!paths.empty()) {
+		bool did_something = false;
+		log("Applying %s at %d top-level path%s.\n",
+		    desc.c_str(), GetSize(paths), paths.size() > 1 ? "s" : "");
+		for (const auto &path : paths)
+			did_something |=
+				bind_path(design, counts, path, binding, top_mod);
+		return did_something;
+	}
+
+	// If neither of these worked, search for a module by name. Note that
+	// this takes no account of design hierarchy.
+	std::vector<RTLIL::Module *> mods = binding.find_concrete_module_targets(design);
+	if (!mods.empty()) {
+		// We can add the binding directly to the module, operating in place.
+		bool did_something = false;
+		log("Applying %s to %d module%s.\n",
+		    desc.c_str(), GetSize(mods), mods.size() > 1 ? "s" : "");
+		for (auto mod : mods)
+			did_something |= binding.bind_into(design, *mod);
+		return did_something;
+	}
+
+	std::ostringstream err_msg;
+	err_msg << "Couldn't find a target for " << desc << ".";
+	binding.report_error(err_msg.str());
+
+	// This is unreachable because binding.report_error() is noreturn.
+	// Unfortunately, the [[noreturn]] attribute cannot be relied on for
+	// virtual functions, so we have to return something here.
+	log_assert(0);
+	return false;
+}
+
+// Worker function for get_cell_type_counts. To avoid double counting, we
+// assume that any module that has already been processed has an entry in
+// counts. Top-level modules get a count of zero added just before calling this
+// function on them.
+static void
+get_cell_type_counts_worker(RTLIL::Design &design,
+                            dict<RTLIL::IdString, int> &counts,
+                            RTLIL::Module &mod)
+{
+	for (auto cell : mod.cells()) {
+		// Set counts[cell->type] to 1 if it isn't defined
+		// yet, or increment if it is.
+		counts[cell->type]++;
+
+		// If the count is greater than 1, we have visited
+		// this cell type before, so we don't need to recurse.
+		if (counts[cell->type] > 1)
+			continue;
+
+		RTLIL::Module *child_mod = design.module(cell->type);
+		if (child_mod)
+			get_cell_type_counts_worker(design,
+						    counts, *child_mod);
+	}
+}
+
+// Count the number of times cell type in the design appears as a cell in
+// another module. Cell types for modules that only appear at top-level have a
+// count of zero.
+//
+// Returns a dictionary keyed by module name. Every module will appear in the
+// dictionary, plus any other cell that's used.
+static dict<RTLIL::IdString, int>
+get_cell_type_counts(RTLIL::Design &design)
+{
+	dict<RTLIL::IdString, int> counts;
+	for (auto mod : design.modules()) {
+		counts.insert(std::make_pair(mod->name, 0));
+		get_cell_type_counts_worker(design, counts, *mod);
+	}
```

Comment:
After poking around some more, it seems this procedure can double count. Running `hierarchy -top c` with the below modules loaded, the count of `a` internally becomes 2 because `b` is itself visited twice.

```SystemVerilog
module m;
endmodule
module a;
endmodule
module b;
    a x();
    bind x m z();
endmodule
module c;
    b y();
endmodule
```